From 1d1751fbbaf6ae74c991812e2a172c88c356e1ee Mon Sep 17 00:00:00 2001
From: Qinxin Xia <xiaqinxin@huawei.com>
Date: Thu, 28 Aug 2025 17:05:20 +0800
Subject: [PATCH 756/960] downstream: gic: increase the arch_timer priority to
 avoid hardlockup

The current interrupt policy of the Hip12 preferentially
schedules LPI and SGI interrupts with the same priority.
As a result, the arch_timer may not be executed when frequent
LPI/SGI interrupts are generated, and the watchdog is starved.

Signed-off-by: Qinxin Xia <xiaqinxin@huawei.com>
Signed-off-by: Hongye Lin <linhongye@h-partners.com>
Signed-off-by: Slim6882 <yangjunshuo@huawei.com>

 Conflicts:
	arch/arm64/Kconfig
	config.aarch64
	drivers/clocksource/arm_arch_timer.c
	include/linux/irqchip/arm-gic-v3.h

Signed-off-by: Slim6882 <yangjunshuo@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/Kconfig                   | 10 +++++++
 config.aarch64                       |  1 +
 config.aarch64-64k                   |  1 +
 drivers/clocksource/arm_arch_timer.c | 45 ++++++++++++++++++++++++++++
 drivers/irqchip/irq-gic-v3.c         |  2 +-
 include/linux/irqchip/arm-gic-v3.h   |  4 +++
 6 files changed, 62 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index b0cae87f4c1b..740077d82d7f 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -1267,6 +1267,16 @@ config HISILICON_ERRATUM_162100125
 	  CNP support for Hisilicon LINXICORE9100 cores.
 
 	  If unsure, say Y.
+config HISILICON_ERRATUM_165010801
+	bool "Hisilicon erratum 165010801"
+	depends on ARCH_HISI
+	default y
+	help
+	  On HIP12, when GIC receives multiple interrupts of the same priority and
+	  different types, the interrupts are selected in the following sequence:
+	  SPI > LPI > SGI > PPI. This scheduling rule may cause PPI starvation.
+	  To prevent starvation from triggering system watchdog hardlockup, the
+	  interrupt priority is explicitly increased in the arch_timer driver.
 
 config HISILICON_ERRATUM_1980005
 	bool "Hisilicon erratum 1980005: IDC support"
diff --git a/config.aarch64 b/config.aarch64
index 8b9ce8777b94..c41059c7f462 100644
--- a/config.aarch64
+++ b/config.aarch64
@@ -8663,6 +8663,7 @@ CONFIG_HISILICON_ERRATUM_161010101=y
 CONFIG_ARM64_ERRATUM_858921=y
 CONFIG_SUN50I_ERRATUM_UNKNOWN1=y
 CONFIG_ARM_TIMER_SP804=y
+CONFIG_HISILICON_ERRATUM_165010801=y
 # end of Clock Source drivers
 
 CONFIG_MAILBOX=y
diff --git a/config.aarch64-64k b/config.aarch64-64k
index d21be40b3358..2ba7f82da2c2 100644
--- a/config.aarch64-64k
+++ b/config.aarch64-64k
@@ -8656,6 +8656,7 @@ CONFIG_HISILICON_ERRATUM_161010101=y
 CONFIG_ARM64_ERRATUM_858921=y
 CONFIG_SUN50I_ERRATUM_UNKNOWN1=y
 CONFIG_ARM_TIMER_SP804=y
+CONFIG_HISILICON_ERRATUM_165010801=y
 # end of Clock Source drivers
 
 CONFIG_MAILBOX=y
diff --git a/drivers/clocksource/arm_arch_timer.c b/drivers/clocksource/arm_arch_timer.c
index e054de92de91..f664f2af3352 100644
--- a/drivers/clocksource/arm_arch_timer.c
+++ b/drivers/clocksource/arm_arch_timer.c
@@ -29,6 +29,11 @@
 #include <linux/arm-smccc.h>
 #include <linux/ptp_kvm.h>
 
+#ifdef CONFIG_HISILICON_ERRATUM_165010801
+#include <linux/irqchip/arm-gic-common.h>
+#include <linux/irqchip/arm-gic-v3.h>
+#endif
+
 #include <asm/arch_timer.h>
 #include <asm/virt.h>
 
@@ -93,6 +98,10 @@ static enum vdso_clock_mode vdso_default = VDSO_CLOCKMODE_ARCHTIMER;
 static enum vdso_clock_mode vdso_default = VDSO_CLOCKMODE_NONE;
 #endif /* CONFIG_GENERIC_GETTIMEOFDAY */
 
+#ifdef CONFIG_HISILICON_ERRATUM_165010801
+static bool prio_setup;
+#endif
+
 static cpumask_t evtstrm_available = CPU_MASK_NONE;
 static bool evtstrm_enable __ro_after_init = IS_ENABLED(CONFIG_ARM_ARCH_TIMER_EVTSTREAM);
 
@@ -355,6 +364,17 @@ static struct ate_acpi_oem_info hisi_161010101_oem_info[] = {
 };
 #endif
 
+#ifdef CONFIG_HISILICON_ERRATUM_165010801
+static struct ate_acpi_oem_info hisi_165010801_oem_info[] = {
+	{
+		.oem_id		= "HISI  ",
+		.oem_table_id	= "HIP12   ",
+		.oem_revision	= 0,
+	},
+	{ /* Sentinel indicating the end of the OEM array */ },
+};
+#endif
+
 #ifdef CONFIG_ARM64_ERRATUM_858921
 static u64 notrace arm64_858921_read_cntpct_el0(void)
 {
@@ -485,6 +505,13 @@ static const struct arch_timer_erratum_workaround ool_workarounds[] = {
 		.set_next_event_virt = erratum_set_next_event_virt,
 	},
 #endif
+#ifdef CONFIG_HISILICON_ERRATUM_165010801
+	{
+		.match_type = ate_match_acpi_oem_info,
+		.id = hisi_165010801_oem_info,
+		.desc = "HiSilicon erratum 165010801",
+	},
+#endif
 #ifdef CONFIG_ARM64_ERRATUM_858921
 	{
 		.match_type = ate_match_local_cap_id,
@@ -592,6 +619,12 @@ void arch_timer_enable_workaround(const struct arch_timer_erratum_workaround *wa
 	if (wa->read_cntvct_el0 || wa->read_cntpct_el0)
 		atomic_set(&timer_unstable_counter_workaround_in_use, 1);
 
+#ifdef CONFIG_HISILICON_ERRATUM_165010801
+	if (!strncmp(wa->desc, "HiSilicon erratum 165010801",
+		     strlen("HiSilicon erratum 165010801")))
+		prio_setup = true;
+#endif
+
 	/*
 	 * Don't use the vdso fastpath if errata require using the
 	 * out-of-line counter accessor. We may change our mind pretty
@@ -1031,6 +1064,18 @@ static int arch_timer_starting_cpu(unsigned int cpu)
 
 	__arch_timer_setup(ARCH_TIMER_TYPE_CP15, clk);
 
+#ifdef CONFIG_HISILICON_ERRATUM_165010801
+	if (prio_setup && is_kernel_in_hyp_mode()) {
+		struct irq_data *d = irq_get_irq_data(arch_timer_ppi[arch_timer_uses_ppi]);
+
+		if (!d)
+			pr_warn("WARNING: Invalid arch_timer ppi irq: %d!\n",
+				arch_timer_ppi[arch_timer_uses_ppi]);
+		else
+			gic_irq_set_prio(d, GICD_INT_DEF_PRI & (GICD_INT_DEF_PRI - 1));
+	}
+#endif
+
 	flags = check_ppi_trigger(arch_timer_ppi[arch_timer_uses_ppi]);
 	enable_percpu_irq(arch_timer_ppi[arch_timer_uses_ppi], flags);
 
diff --git a/drivers/irqchip/irq-gic-v3.c b/drivers/irqchip/irq-gic-v3.c
index 249c012b1790..635a4169e11d 100644
--- a/drivers/irqchip/irq-gic-v3.c
+++ b/drivers/irqchip/irq-gic-v3.c
@@ -541,7 +541,7 @@ static int gic_irq_get_irqchip_state(struct irq_data *d,
 	return 0;
 }
 
-static void gic_irq_set_prio(struct irq_data *d, u8 prio)
+void gic_irq_set_prio(struct irq_data *d, u8 prio)
 {
 	void __iomem *base = gic_dist_base(d);
 	u32 offset, index;
diff --git a/include/linux/irqchip/arm-gic-v3.h b/include/linux/irqchip/arm-gic-v3.h
index ae139ff41f0a..826e8b96d19e 100644
--- a/include/linux/irqchip/arm-gic-v3.h
+++ b/include/linux/irqchip/arm-gic-v3.h
@@ -756,6 +756,10 @@ static inline enum gic_intid_range __get_intid_range(irq_hw_number_t hwirq)
 	}
 }
 
+
+#ifdef CONFIG_HISILICON_ERRATUM_165010801
+extern void gic_irq_set_prio(struct irq_data *d, u8 prio);
+#endif
 #endif
 
 #endif
-- 
2.43.0

